### PR TITLE
Add tiling module and CLI options

### PR DESCRIPTION
**Pull request by clausnagel**
This PR adds a tiling module and API as well as corresponding tiling options to the CLI command.

### CODE REVIEW

**Comment by yaozhihang** [comment]
The following error was produced while running the import/export CLI command. The same command works well with the main branch. Is there anything I missed?

```
[22:27:26 ERROR] Failed to connect to the database.
[22:27:26 ERROR] org.citydb.database.DatabaseException: No database adapter available for the database 'PostgreSQL'.
[22:27:26 WARN] citydb execution failed.
````

**Comment by clausnagel** [comment]
Hm, cannot reproduce. Maybe you can debug?

**Comment by clausnagel** [comment]
Test again. I fixed the dependencies.

**Comment by yaozhihang** [comment]
Works now. What should the json config look like? for example in case of matrix tiling like this?
```
"tiling": {
  "schema": {
    "columns": 2,
    "rows": 2
    }
 }
```

File: citydb-tiling/src/main/java/org/citydb/tiling/options/DimensionScheme.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+import java.sql.SQLException;
+import java.util.Optional;
+
+@JSONType(typeName = "Dimension")
+public class DimensionScheme extends TilingScheme {
+    private Dimension width;
+    private Dimension height;
+    @JSONField(serializeUsing = PointWriter.class, deserializeUsing = PointReader.class)
+    private Point gridPoint;
```
Comment (yaozhihang) [inline citydb-tiling/src/main/java/org/citydb/tiling/options/DimensionScheme.java:55]: The gridPoint parameter not available in CLI?

**Comment by yaozhihang** [comment]
The `TileMatrix` is always based on the database srs. Shouldn't the matrix based on the input SRS (e.g., WGS84) of the DimensionSchema?

**Review by yaozhihang — COMMENTED** [review]
Very nice PR. One important remark is about the CRS of the TileMatrix.

File: citydb-tiling/src/main/java/org/citydb/tiling/options/DimensionScheme.java (hunk 1 above)
Comment (clausnagel) [reply citydb-tiling/src/main/java/org/citydb/tiling/options/DimensionScheme.java:55]: No

**Comment by clausnagel** [comment]
> Works now. What should the json config look like? for example in case of matrix tiling like this?

You have to add the `type` attribute to identify the tiling scheme. For example:

```json
"tiling":{
	"scheme":{
		"type":"Dimension",
		"height":{
			"value":1000.0
		},
		"width":{
			"unit":"km",
			"value":1.0
		}
	}
}
```

or 

```json
"scheme":{
	"type":"Matrix",
	"columns":4,
	"rows":4
}
```

**Comment by clausnagel** [comment]
> The `TileMatrix` is always based on the database srs. Shouldn't the matrix based on the input SRS (e.g., WGS84) of the DimensionSchema?

The TileMatrix produces the tiles which are used as spatial filter against the database. I think this should be done in the CRS of the database to avoid distortions. What would be the advantage of doing this in WGS84?

File: citydb-tiling/src/main/java/org/citydb/tiling/options/DimensionScheme.java (hunk 1 above)
Comment (clausnagel) [reply citydb-tiling/src/main/java/org/citydb/tiling/options/DimensionScheme.java:55]: Additional information: When using the `Dimension` tiling scheme, the tiles are automatically aligned with the database CRS. For example, when using 2x2km as width and height, the lower-left x-values of the tiles will be ..., -2000, 0, 2000, 4000, 6000, 8000, ... (and same for the y-values in this example). Only those tiles will be exported which cover the user-defined or auto-computed extent of the requested features, of course.

The `gridPoint` attribute allows a user to define a fixed point for tile alignment. For instance, if the user defines `(750,750)` as grid point in the above example, the the lower-left x-values would be ..., -1250, 750, 2750, 4750, ...

**Comment by yaozhihang** [comment]
> > The `TileMatrix` is always based on the database srs. Shouldn't the matrix based on the input SRS (e.g., WGS84) of the DimensionSchema?
> 
> The TileMatrix produces the tiles which are used as spatial filter against the database. I think this should be done in the CRS of the database to avoid distortions. What would be the advantage of doing this in WGS84?

Tiling based on the input SRS can provide a more generic way.  One use case is the current VIS-Exporter, which ist based on the tiling in WGS84. 

Another case is for example the CRS used  by the iUR data, which is a Geographic CRS. Tiling based an input projected CRS can avoid distortions.

File: citydb-tiling/src/main/java/org/citydb/tiling/options/DimensionScheme.java (hunk 1 above)
Comment (yaozhihang) [reply citydb-tiling/src/main/java/org/citydb/tiling/options/DimensionScheme.java:55]: OK. seems not very necessary to add it in CLI.

**Comment by yaozhihang** [comment]
> > Works now. What should the json config look like? for example in case of matrix tiling like this?
> 
> You have to add the `type` attribute to identify the tiling scheme. And be careful to use `scheme` not `schema`. For example:
> 
> ```json
> "tiling":{
> 	"scheme":{
> 		"type":"Dimension",
> 		"height":{
> 			"value":1000.0
> 		},
> 		"width":{
> 			"unit":"km",
> 			"value":1.0
> 		}
> 	}
> }
> ```
> 
> or
> 
> ```json
> "tiling":{
> 	"scheme":{
> 		"type":"Matrix",
> 		"columns":4,
> 		"rows":4
> 	}
> }
> ```

cool, works well.

**Comment by clausnagel** [comment]
> > > The `TileMatrix` is always based on the database srs. Shouldn't the matrix based on the input SRS (e.g., WGS84) of the DimensionSchema?
> > 
> > 
> > The TileMatrix produces the tiles which are used as spatial filter against the database. I think this should be done in the CRS of the database to avoid distortions. What would be the advantage of doing this in WGS84?
> 
> Tiling based on the input SRS can provide a more generic way. One use case is the current VIS-Exporter, which ist based on the tiling in WGS84.
> 
> Another case is for example the CRS used by the iUR data, which is a Geographic CRS. Tiling based an input projected CRS can avoid distortions.

Tiling works in case the user provides the tiling extent in the database CRS or in any other CRS. The database CRS is only used internally for computing the tiles. These are the steps of the VIS export implementation of the old Importer/Exporter:

1. The user provides the tiling extent in the GUI in any CRS, e.g. WGS84
2. The VIS exporter transforms the extent into the database CRS to calculate the tile matrix (columns, rows)
3. The VIS exporter transforms the tile extent back to WGS84
4. The VIS exporter iterates over all tiles. Because of step 3, the extent of each tile is also in WGS84
5. To use the tile as spatial filter, the VIS exporter transforms the extent of each tile back to the database CRS
6. The VIS exporter receives features from the database in the database CRS. To check whether a feature is on the tile, the center point of the feature's envelope is used. Since this center point is in the database CRS but the tile extent in WGS84, the center point must again be transformed to WGS84 to perform the check.

I think this is not a good example. The new API is much clearer in my opinion.

**Comment by yaozhihang** [comment]
The new API is clearer. Just hope it to be more generic to support tiling matrix with different CRS. like the old API
https://github.com/3dcitydb/importer-exporter/blob/master/impexp-core/src/main/java/org/citydb/core/query/filter/tiling/Tiling.java#L90

**Review by yaozhihang — APPROVED** [review]

**Comment by clausnagel** [comment]
Merging now. The extents of both the `TileMatrix` and of individual tiles can be transformed to any other CRS using the `transform` method of `GeometryAdapter`.

**Pull request merged**